ethereum/core/src/test/java/tech/pegasys/teku/core/signatures/DeletableSignerTest.java

```diff
@@ -0,0 +1,230 @@
+        .isCompletedExceptionallyWith(SignerNotActiveException.class);
+    verify(delegate, never()).signSyncCommitteeMessage(UInt64.ONE, root, forkInfo);
+  }
+
+  @Test
+  void signSyncCommitteeSelectionProof_shouldAlwaysSign() {
+    final SyncAggregatorSelectionData syncAggregatorSelectionData =
+        syncCommitteeUtil.createSyncAggregatorSelectionData(UInt64.ONE, UInt64.valueOf(1));
+    when(delegate.signSyncCommitteeSelectionProof(syncAggregatorSelectionData, forkInfo))
+        .thenReturn(signatureFuture);
+
+    assertThatSafeFuture(
+            signer.signSyncCommitteeSelectionProof(syncAggregatorSelectionData, forkInfo))
+        .isCompletedWithValue(signature);
+  }
+
+  @Test
+  void signSyncCommitteeSelectionProof_shouldNotSignWhenDisabled() {
+    final SyncAggregatorSelectionData syncAggregatorSelectionData =
+        syncCommitteeUtil.createSyncAggregatorSelectionData(UInt64.ONE, UInt64.valueOf(1));
+    when(delegate.signSyncCommitteeSelectionProof(syncAggregatorSelectionData, forkInfo))
+        .thenReturn(signatureFuture);
+
+    assertThatSafeFuture(
+            signer.signSyncCommitteeSelectionProof(syncAggregatorSelectionData, forkInfo))
+        .isCompletedWithValue(signature);
+  }
+
+  // signContributionAndProof
+  @Test
+  void signContributionAndProof_shouldAlwaysSign() {
+
+    final ContributionAndProof contributionAndProof =
+        dataStructureUtil.randomSignedContributionAndProof(6L).getMessage();
+    when(delegate.signContributionAndProof(contributionAndProof, forkInfo))
+        .thenReturn(signatureFuture);
+
+    assertThatSafeFuture(signer.signContributionAndProof(contributionAndProof, forkInfo))
+        .isCompletedWithValue(signature);
+  }
+
+  @Test
+  void signContributionAndProof_shouldNotSignWhenDisabled() {
+    final ContributionAndProof contributionAndProof =
+        dataStructureUtil.randomSignedContributionAndProof(6L).getMessage();
+    when(delegate.signContributionAndProof(contributionAndProof, forkInfo))
+        .thenReturn(signatureFuture);
+
+    assertThatSafeFuture(signer.signContributionAndProof(contributionAndProof, forkInfo))
+        .isCompletedWithValue(signature);
+  }
+
+  @Test
+  void shouldNotMarkDeletedWhileSigning() {
+    SafeFuture<BLSSignature> future = new SafeFuture<>();
+    final BeaconBlock block = dataStructureUtil.randomBeaconBlock(6);
+    when(delegate.signBlock(block, forkInfo)).thenReturn(future);
+
+    final SafeFuture<BLSSignature> future2 = signer.signBlock(block, forkInfo);
+    assertThatSafeFuture(future2).isNotCompleted();
+    final SafeFuture<Void> disabledSigner =
+        SafeFuture.of(SafeFuture.runAsync(() -> signer.disable()));
+
+    assertThatSafeFuture(disabledSigner).isNotCompleted();
+    future.complete(dataStructureUtil.randomSignature());
+    assertThatSafeFuture(future2).isCompleted();
+    assertThat(disabledSigner).isCompleted();
+  }
```

Comment:
There's a race condition here where the thread calling `signer.disable()` may not hav even started by the time we check `disabledSigner.isNotCompleted()` causing the test to fail even if there was no thread safety.  It's more complex but I think in this case it's worth actually waiting on the appropriate thread states to guarantee we're testing the locking.

```
  @Test
  void shouldNotMarkDeletedWhileSigning() {
    SafeFuture<BLSSignature> future = new SafeFuture<>();
    final BeaconBlock block = dataStructureUtil.randomBeaconBlock(6);
    when(delegate.signBlock(block, forkInfo)).thenReturn(future);

    final SafeFuture<BLSSignature> future2 = signer.signBlock(block, forkInfo);
    assertThatSafeFuture(future2).isNotCompleted();
    final SafeFuture<Void> disablingComplete = new SafeFuture<>();
    final Thread disableThread =
        new Thread(
            () -> {
              signer.disable();
              disablingComplete.complete(null);
            });
    disableThread.start();

    // Wait for the signer.disable() call to be waiting for the write lock
    waitForThreadState(disableThread, State.WAITING);

    assertThatSafeFuture(disablingComplete).isNotCompleted();
    future.complete(dataStructureUtil.randomSignature());
    assertThatSafeFuture(future2).isCompleted();

    // Confirm the disable can complete once signing has completed.
    waitForThreadState(disableThread, State.TERMINATED);
    assertThat(disablingComplete).isCompleted();
  }

  private void waitForThreadState(final Thread thread, final State state) {
    final long startTime = System.currentTimeMillis();
    while (thread.getState() != state) {
      if (System.currentTimeMillis() - startTime > 5000) {
        Assertions.fail("Thread did not reach state " + state + " in a reasonable time");
      }
      LockSupport.parkNanos(TimeUnit.MILLISECONDS.toNanos(500));
    }
  }
```

Reply:
Might be worth having another variant of that test where the signing actually throws an exception rather than completing successfully and it should still release the lock and let the disable method complete.

Reply:
i do like that test